enhancements/cvo/clusteroperator-resource-handling.md

```diff
@@ -0,0 +1,117 @@
+last-updated: 2019-12-11
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# ClusterOperator Resource Handling
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+`clusteroperators.config.openshift.io` are used to determine success of installation.
+They are also used to drive collection of debugging data from tools like `oc adm inspect` and `oc adm must-gather`.
+The `clusteroperator` resources in a payload should always be present, even before installation of particular
+operators to see which `clusteroperators` need to report in and to allow establishing `.status.relatedResources` before
+an operator pod runs.
+This is critical for debugging clusters that fail to install or fail to upgrade with new operators present.
+
+## Motivation
+
+Debugging information for clusters that succeed in bootstrapping, but fail during installation is missing most
+of the data required to resolve the issue via must-gather.
+
+### Goals
+
+1. allow collection of debugging data for failed installs using normal tools. 
+
+### Non-Goals
+
+1. create a new tool to gather data for failed installs after bootstrapping.
+2. taking responsibility for creating clusteroperator resources.  Individual operators are responsible for creating and maintaining
+   clusteroperator resources.
+
+## Proposal
+
+1. `clusteroperator` resources in the payload should be created with the required status conditions (available, progressing,
+   degraded) set to `Unknown`.
+2. `clusteroperator` creation by the CVO needs to honor or update `.status.relatedResources`.  This requires updating
+    status after the creation.
+3.  `clusteroperator` resources in the payload should all be created immediately regardless of where in the payload ordering
+    they are located.  This ensures that they are always present during collection.
+4.  The CVO waiting logic on `clusteroperator` remains the same.
+
+### Specific Implementation Option
+
+This isn't a required mechanism for implementation, but it demonstrates how narrowly scoped the change is.
+ 1. create a new control loop with a clusteroperator lister, clusteroperator client, and a function to get the current payload.
+ 2. register event handlers on clusteroperator informer and time based every minute.
+ 3. sync loop reads the current payload.  for each clusteroperator in the payload
+    1. check lister to see if clusteroperator exists.  If so, continue to next clusteroperator.
+    2. create clusteroperator with empty spec and metadata.  If create fails, continue to next clusteroperator.
+    3. update clusteroperator/status with `.status.relatedResources` and the three required conditions in `Unknown` state.
```

Comment:
What needs to go in `relatedResources`?

Reply:
> What needs to go in `relatedResources`?

Responsibility of the individual operators.  Generally it will be things like input resources with fixed names and interesting namespaces.

Reply:
so as far as this enhancement is concerned, we can drop that and have this line (for the CVO) be:

> 3. update clusteroperator/status with the three required conditions in `Unknown` state.

Reply:
> so as far as this enhancement is concerned, we can drop that...

And without anything useful in `relatedObjects` (is `relatedResources` a typo?), must-gather is still stuck, right?  I don't see how the ClusterVersion operator would know what to fill in there unless we put some sort of annotation on the intended, CVO created resources.  So an operator's manifest set would look like:

1. Namespace, whatever else we expect to not have problems with, all have the `release.openshift.io/operator-resource: <operator-name>` annotation.
2. The ClusterOperator.  When the CVO sees this is missing, it creates it, seeds the conditions as you've laid out, and fills `relatedObjects` with anything it had already pushed that sync round with the `release.openshift.io/operator-resource` annotation whose annotation value matched the ClusterOperator name.
3. Deployments or other vulnerable types.  At this point it would be an error to hit anything with a matching `release.openshift.io/operator-resource` annotation.  Not sure how to enforce that error, probably go Degraded but keep working.

Reply:
>  So an operator's manifest set would look like...

Never mind, this is crazy.  Better to have operator maintainers set their intended `relatedObjects` content in their ClusterOperator.